### DISCUSSION TITLE
Request does not raise UnsupportedMediaType exception

### DISCUSSION THREAD

**Discussion by Feverfew** [encode/django-rest-framework#7820] [General]
## Checklist
- [x] I have reduced the issue to the simplest possible case.

With the following view:
```python
@api_view(['POST'])
@parser_classes([JSONParser])
def my_view(request):
    request.data['change_me'] = 'MyValue'
```
By sending a request with `Content-Type: multipart/form-data` (with at least one key/value pair, if it's empty it still goes through but doesn't trip up weirdly enough) one will observe the following behaviour:
`AttributeError at $myViewsURL This QueryDict instance is immutable`

What surprises me is that as I've specified the use of only the `JSONParser`, I should expect `request.data` to be a `dict`. I've tried other content-types and they are correctly rejected before parsing, but this one goes through.

**Comment by lovelydinosaur** [comment]
Guiding you kindly towards discussions... :)

<img width="743" alt="Screenshot 2021-03-10 at 09 22 38" src="https://user-images.githubusercontent.com/647359/110606647-4cd9b600-8182-11eb-91e2-43787a5f279e.png">

I'm not able to replicate this...

**urls.py**

```python
from django.urls import path
from rest_framework.decorators import api_view, parser_classes
from rest_framework.parsers import JSONParser


@api_view(['POST'])
@parser_classes([JSONParser])
def homepage(request):
    request.data['change_me'] = 'MyValue'


urlpatterns = [
    path('', homepage),
]
```

**client.py**:

```
import httpx

r = httpx.post("http://127.0.0.1:8000", data={"a":"b"})
print(r)
```

**shell**:

```
$ python client.py 
<Response [415 Unsupported Media Type]>
```

**Reply by Feverfew** [reply]
I'm able to replicate this with the following (i.e. make sure `Content-Type == 'multipart/form-data'`, your test case likely sets it to `application/x-www-form-urlencoded`):
```python
>>> import requests
>>> r = requests.post("http://127.0.0.1:8000", files={"a":"b"})
>>> print(r)
<Response [500]>
```

The issues occurs in `request.py`, in particular in `def _parse(self)`. IMO this should also return a `415` before it gets to my code in the view.

**Reply by Feverfew** [reply]
Further debugging indicates that in `def _parse(self)` a `RawPostDataException` is raised (which is caught appropriately). Interestingly in `tests/test_response.py` a similar test (`UnsupportedMediaTypeExists.test_should_not_allow_posting_a_form`) already exists but the exception isn't raised, leading to the expected behaviour.

**Reply by Feverfew** [reply]
Ok, I believe I have figured out the issue. This is not a bug in Django Rest Framework but actually a bug in my code!

For debugging purposes we have a special [middleware](https://github.com/uclapi/uclapi/blob/c5999617a02e4c0d00f8120794884a56c6428a86/backend/uclapi/dashboard/middleware/fake_shibboleth_middleware.py) that we use, which modifies/accesses `request.POST`, which is explicitly [prohibited ](https://stackoverflow.com/a/28641930)and is why the `RawPostDataException` is raised.

**Reply by lovelydinosaur** [reply]
Fanastic, thanks for following up on that.
